viewing_party/party.py

```diff
@@ -1,23 +1,183 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+
+    new_movie = {}
+
+    # check that all necessary movie characteristics exist before creating new movie
+    if title and genre and rating:
+        new_movie["title"] = title
+        new_movie["genre"] = genre
+        new_movie["rating"] = rating
+    else:
+        new_movie = None
+
+    return new_movie
+
+def add_to_watched(user_data, movie):
+
+    # add movie to watched list
+    user_data["watched"].append(movie)
+
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+
+    # add movie to watchlist
+    user_data["watchlist"].append(movie)
+
+    return user_data
+
+def watch_movie(user_data, movie_title):
+
+    # move movies that user watches from watchlist to watched
+    for movie in user_data["watchlist"]:
+        if movie["title"] == movie_title:
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)
+    
+    return user_data

```

Comment:
This function works, but as an aside it contains an "anti-pattern" that is important to be aware of. An element of the list that is being iterated over is being removed inside the for loop. This won't cause problems in this particular case, but consider this slight change:
```
def watch_movie(user_data, title):
    for i in range(len(user_data["watchlist"])):
        item = user_data["watchlist"][i]
        if item["title"] == title:
            user_data["watched"].append(item)
            user_data["watchlist"].remove(item)
    return user_data
```
If we were to test this in a scenario where `user_data["watchlist"]` has 2 elements, and the first is the movie that we are moving to `user_data["watched"]`, the second iteration through the loop will throw an index error, because it will try to access the 2nd element in `user_data["watchlist"]`, but there no longer is a second element.  

As a coding pattern, this is one I recommend avoiding, because it has so much potential for unforeseen errors. One fix is to return immediately after making the swap (ie, move line 39 inside the if-block so that the loop ends as soon as the swap is made). Another fix would be to use the loop to find the item, but not make the swap until after the loop ends.

From a performance/time standpoint, consider if we were dealing with a list of 1,000 movies. If the item we wanted to swap happens to be the first item, we would continue looping over 999 titles before the loop exits. Exiting early also helps us avoid extra time and calculations that don't add to our final result.

Reply:
Yes, I'd love to talk through the danger zone piece in our one on one. My gut reaction would be to create a copy of user_data["watchlist"] and remove movies from that copy rather than the original list. But again, I'd love to talk through this with you.

I also really appreciate the note about time complexity! My first guess would be to add a break on line 38 within the conditional so that the for loop would stop as soon as a match was found and the code could skip ahead to the return statement.